File: services/backend/pkg/gamification/gamification_test.go
package gamification

import "testing"

func TestXPForSession_BaseDistance(t *testing.T) {
	// 5 km, < 10 km, no hr → 5 XP base.
	xp := XPForSession(SessionInput{DistanceM: 5000, DurationS: 1500})
	if xp != 5 {
		t.Errorf("expected 5 xp, got %d", xp)
	}
}

func TestXPForSession_LongRunBonus(t *testing.T) {
	// 10 km exactly → 10 km + 5 bonus = 15.
	xp := XPForSession(SessionInput{DistanceM: 10000, DurationS: 3000})
	if xp != 15 {
		t.Errorf("expected 15 xp (10 + bonus), got %d", xp)
	}
	// 16 km → 16 + 5 = 21.
	xp = XPForSession(SessionInput{DistanceM: 16000, DurationS: 4500})
	if xp != 21 {
		t.Errorf("expected 21 xp, got %d", xp)
	}
	// 9.999 km → 9 (no bonus).
	xp = XPForSession(SessionInput{DistanceM: 9999, DurationS: 3000})
	if xp != 9 {
		t.Errorf("expected 9 xp (no long bonus), got %d", xp)
	}
}

func TestXPForSession_HRBonus(t *testing.T) {
	// 5 km + avgHr 155 → 5 + 3 = 8.
	xp := XPForSession(SessionInput{DistanceM: 5000, DurationS: 1500, AvgHrBpm: 155})
	if xp != 8 {
		t.Errorf("expected 8 xp (5 + hr bonus), got %d", xp)
	}
	// avg 149 — без бонуса.
	xp = XPForSession(SessionInput{DistanceM: 5000, DurationS: 1500, AvgHrBpm: 149})
	if xp != 5 {
		t.Errorf("expected 5 xp (below hr threshold), got %d", xp)
	}
}

func TestXPForSession_AllBonuses(t *testing.T) {
	// 16 km + avgHr 160 → 16 + 5 (long) + 3 (hr) = 24.
	xp := XPForSession(SessionInput{DistanceM: 16000, DurationS: 4500, AvgHrBpm: 160})
	if xp != 24 {
		t.Errorf("expected 24 xp, got %d", xp)
	}
}

func TestXPForSession_ZeroInputs(t *testing.T) {
	if XPForSession(SessionInput{}) != 0 {
		t.Error("zero input should give 0 xp")
	}
	if XPForSession(SessionInput{DistanceM: 1000}) != 0 {
		t.Error("zero duration should give 0 xp")
	}
	if XPForSession(SessionInput{DurationS: 100}) != 0 {
		t.Error("zero distance should give 0 xp")
	}
}

func TestGradeForXP(t *testing.T) {
	cases := []struct {
		xp    int
		grade string
	}{
		{0, "D"}, {50, "D"}, {99, "D"},
		{100, "D+"}, {249, "D+"},
		{250, "C"}, {499, "C"},
		{500, "C+"}, {999, "C+"},
		{1000, "B"}, {1999, "B"},
		{2000, "B+"}, {3499, "B+"},
		{3500, "A"}, {5499, "A"},
		{5500, "A+"}, {9999, "A+"},
		{10000, "S"}, {100000, "S"},
	}
	for _, c := range cases {
		got := GradeForXP(c.xp)
		if got != c.grade {
			t.Errorf("xp=%d: expected %s, got %s", c.xp, c.grade, got)
		}
	}
}

func TestGradeForXP_Negative(t *testing.T) {
	if GradeForXP(-100) != "D" {
		t.Error("negative xp should fall back to D")
	}
}

func TestNextGrade(t *testing.T) {
	// 0 XP → next D+ at 100 (100 remaining).
	g, r := NextGrade(0)
	if g != "D+" || r != 100 {
		t.Errorf("expected D+ / 100, got %s / %d", g, r)
	}
	// 750 → next B at 1000 (250 remaining).
	g, r = NextGrade(750)
	if g != "B" || r != 250 {
		t.Errorf("expected B / 250, got %s / %d", g, r)
	}
	// 10000 → at S, no next.
	g, r = NextGrade(10000)
	if g != "" || r != 0 {
		t.Errorf("expected ''/0 at max, got %s / %d", g, r)
	}
}

func TestFormulaVersion(t *testing.T) {
	if FormulaVersion != 1 {
		t.Errorf("formula version drift: %d", FormulaVersion)
	}
}

File: apps/mobile-rn/src/__tests__/gamification.test.ts
// modules/gamification mirror tests — должны 1-к-1 совпадать с Go
// pkg/gamification/gamification_test.go. Расхождение → fail обеих сторон.
//
// Phase 8 / M3.

import { gradeForXP, nextGrade } from '../modules/gamification/domain/grade';
import { FORMULA_VERSION, xpForSession } from '../modules/gamification/domain/xp';

describe('xpForSession base', () => {
  it('5 km / no hr → 5 xp', () => {
    expect(xpForSession({ distanceM: 5000, durationS: 1500 })).toBe(5);
  });
  it('zero distance → 0', () => {
    expect(xpForSession({ distanceM: 0, durationS: 1500 })).toBe(0);
  });
  it('zero duration → 0', () => {
    expect(xpForSession({ distanceM: 5000, durationS: 0 })).toBe(0);
  });
  it('floor partial km', () => {
    expect(xpForSession({ distanceM: 5999, durationS: 1500 })).toBe(5);
  });
});

describe('xpForSession long-run bonus', () => {
  it('10 km → 10 + 5 = 15', () => {
    expect(xpForSession({ distanceM: 10000, durationS: 3000 })).toBe(15);
  });
  it('16 km → 16 + 5 = 21', () => {
    expect(xpForSession({ distanceM: 16000, durationS: 4500 })).toBe(21);
  });
  it('9.999 km — no bonus', () => {
    expect(xpForSession({ distanceM: 9999, durationS: 3000 })).toBe(9);
  });
});

describe('xpForSession HR bonus', () => {
  it('5 km + hr 155 → 5 + 3 = 8', () => {
    expect(xpForSession({ distanceM: 5000, durationS: 1500, avgHrBpm: 155 })).toBe(8);
  });
  it('hr 149 — no bonus', () => {
    expect(xpForSession({ distanceM: 5000, durationS: 1500, avgHrBpm: 149 })).toBe(5);
  });
});

describe('xpForSession all bonuses', () => {
  it('16 km + hr 160 → 16 + 5 + 3 = 24', () => {
    expect(xpForSession({ distanceM: 16000, durationS: 4500, avgHrBpm: 160 })).toBe(24);
  });
});

describe('gradeForXP', () => {
  const cases: Array<[number, string]> = [
    [0, 'D'], [99, 'D'],
    [100, 'D+'], [249, 'D+'],
    [250, 'C'], [499, 'C'],
    [500, 'C+'], [999, 'C+'],
    [1000, 'B'], [1999, 'B'],
    [2000, 'B+'], [3499, 'B+'],
    [3500, 'A'], [5499, 'A'],
    [5500, 'A+'], [9999, 'A+'],
    [10000, 'S'], [100000, 'S'],
  ];
  it.each(cases)('xp=%i → %s', (xp, grade) => {
    expect(gradeForXP(xp)).toBe(grade);
  });
  it('negative → D', () => {
    expect(gradeForXP(-100)).toBe('D');
  });
  it('NaN → D', () => {
    expect(gradeForXP(NaN)).toBe('D');
  });
});

describe('nextGrade', () => {
  it('0 xp → D+ at 100 (100 remaining)', () => {
    expect(nextGrade(0)).toEqual({ label: 'D+', xpRemaining: 100 });
  });
  it('750 xp → B at 1000 (250 remaining)', () => {
    expect(nextGrade(750)).toEqual({ label: 'B', xpRemaining: 250 });
  });
  it('at S (10000+) → null', () => {
    expect(nextGrade(10000)).toBeNull();
    expect(nextGrade(50000)).toBeNull();
  });
});

describe('formula version', () => {
  it('aligned with Go pkg/gamification.FormulaVersion', () => {
    expect(FORMULA_VERSION).toBe(1);
  });
});
